File: tests/usecase/user/delete_test.rs
use std::sync::Arc;

use async_trait::async_trait;

use go_clean_template::application::usecase::user::{DeleteUserInput, DeleteUserUseCase, DeleteUserUseCaseImpl, UserRepository};
use go_clean_template::domain;

struct MockUserRepo {
    find_by_id: Arc<std::sync::Mutex<Option<Result<domain::User, domain::Error>>>>,
    delete: Arc<std::sync::Mutex<Option<Result<(), domain::Error>>>>,
}

impl MockUserRepo {
    fn new() -> Self {
        MockUserRepo {
            find_by_id: Arc::new(std::sync::Mutex::new(None)),
            delete: Arc::new(std::sync::Mutex::new(None)),
        }
    }
}

#[async_trait]
impl UserRepository for MockUserRepo {
    async fn find_by_id(&self, _id: &str) -> Result<domain::User, domain::Error> {
        self.find_by_id.lock().unwrap().clone().unwrap_or(Err(domain::Error::NotFound))
    }
    async fn find_all(&self, _offset: usize, _limit: usize) -> Result<(Vec<domain::User>, i64), domain::Error> {
        Ok((vec![], 0))
    }
    async fn find_by_email(&self, _email: &str) -> Result<domain::User, domain::Error> { Err(domain::Error::NotFound) }
    async fn create(&self, _user: &domain::User) -> Result<(), domain::Error> { Ok(()) }
    async fn update(&self, _user: &domain::User) -> Result<(), domain::Error> { Ok(()) }
    async fn delete(&self, _id: &str) -> Result<(), domain::Error> {
        self.delete.lock().unwrap().clone().unwrap_or(Ok(()))
    }
}

#[tokio::test]
async fn test_delete_user_success() {
    let user = domain::User::new("Alice".into(), "a@b.com".into(), "123".into(), vec![]).unwrap();
    let repo = MockUserRepo::new();
    *repo.find_by_id.lock().unwrap() = Some(Ok(user));
    *repo.delete.lock().unwrap() = Some(Ok(()));

    let uc = DeleteUserUseCaseImpl::new(Arc::new(repo));
    let result = uc.execute(DeleteUserInput { id: "1".into() }).await;
    assert!(result.is_ok());
}

#[tokio::test]
async fn test_delete_user_not_found() {
    let repo = MockUserRepo::new();
    *repo.find_by_id.lock().unwrap() = Some(Err(domain::Error::NotFound));

    let uc = DeleteUserUseCaseImpl::new(Arc::new(repo));
    let result = uc.execute(DeleteUserInput { id: "999".into() }).await;
    assert!(result.is_err());
}

File: tests/application/usecase/user/delete_test.go
package user_test

import (
	"context"
	"testing"

	"github.com/stretchr/testify/assert"

	userusecase "go-clean-template/internal/application/usecase/user"
	"go-clean-template/internal/domain"
)

func TestDelete_Success(t *testing.T) {
	mockRepo := new(MockUserRepo)
	uc := userusecase.NewDeleteUserUseCase(mockRepo)

	user := &domain.User{ID: "1", Name: "Alice", Email: "alice@example.com"}
	mockRepo.On("FindByID", "1").Return(user, nil)
	mockRepo.On("Delete", "1").Return(nil)

	err := uc.Execute(context.Background(), &userusecase.DeleteUserInput{ID: "1"})

	assert.NoError(t, err)
	mockRepo.AssertExpectations(t)
}

func TestDelete_NotFound(t *testing.T) {
	mockRepo := new(MockUserRepo)
	uc := userusecase.NewDeleteUserUseCase(mockRepo)

	mockRepo.On("FindByID", "999").Return(nil, domain.ErrNotFound)

	err := uc.Execute(context.Background(), &userusecase.DeleteUserInput{ID: "999"})

	assert.Error(t, err)
	assert.ErrorIs(t, err, domain.ErrNotFound)
	mockRepo.AssertExpectations(t)
}
